File: chat-uikit/src/main/java/io/agora/chat/uikit/widget/chatrow/EaseChatRowFile.java
package io.agora.chat.uikit.widget.chatrow;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.util.Log;
import android.widget.ImageView;
import android.widget.TextView;


import androidx.annotation.NonNull;

import io.agora.CallBack;
import io.agora.chat.ChatClient;
import io.agora.chat.ChatMessage;
import io.agora.chat.NormalFileMessageBody;
import io.agora.chat.uikit.EaseUIKit;
import io.agora.chat.uikit.R;
import io.agora.chat.uikit.provider.EaseFileIconProvider;
import io.agora.util.TextFormater;

/**
 * file for row
 */
public class EaseChatRowFile extends EaseChatRow {
    private static final String TAG = EaseChatRowFile.class.getSimpleName();
    /**
     * file name
     */
    protected TextView fileNameView;
    /**
     * file's size
     */
	protected TextView fileSizeView;
    /**
     * file state
     */
    protected TextView fileStateView;
    private NormalFileMessageBody fileMessageBody;
    private ImageView ivFileIcon;
    private boolean isDownloading = false;

    public EaseChatRowFile(Context context, boolean isSender) {
        super(context, isSender);
    }

    public EaseChatRowFile(Context context, ChatMessage message, int position, Object adapter) {
        super(context, message, position, adapter);
    }

    @Override
	protected void onInflateView() {
	    inflater.inflate(!showSenderType ? R.layout.ease_row_received_file
                : R.layout.ease_row_sent_file, this);
	}

	@Override
	protected void onFindViewById() {
	    fileNameView = (TextView) findViewById(R.id.tv_file_name);
        fileSizeView = (TextView) findViewById(R.id.tv_file_size);
        fileStateView = (TextView) findViewById(R.id.tv_file_state);
        percentageView = (TextView) findViewById(R.id.percentage);
        ivFileIcon = findViewById(R.id.iv_file_icon);
	}

	@Override
	protected void onSetUpView() {
	    fileMessageBody = (NormalFileMessageBody) message.getBody();
//        Uri filePath = fileMessageBody.getLocalUri();
        if(fileStateView != null) {
            fileStateView.setVisibility(GONE);
        }
        fileNameView.setText(fileMessageBody.getFileName());
        fileSizeView.setText(TextFormater.getDataSize(fileMessageBody.getFileSize()));
        setFileIcon(fileMessageBody.getFileName());
//        if (message.direct() == ChatMessage.Direct.SEND){
//            if (EaseFileUtils.isFileExistByUri(context, filePath)
//                    && message.status() == ChatMessage.Status.SUCCESS) {
//                fileStateView.setText(R.string.ease_have_uploaded);
//            }else {
//                fileStateView.setText("");
//            }
//        }
//        if (message.direct() == ChatMessage.Direct.RECEIVE) {
//            if (EaseFileUtils.isFileExistByUri(context, filePath)) {
//                fileStateView.setText(R.string.ease_have_downloaded);
//            } else {
//                fileStateView.setText(R.string.ease_did_not_download);
//            }
//        }
	}

    @Override
    protected void onMessageSuccess() {
        super.onMessageSuccess();
        if (message.direct() == ChatMessage.Direct.SEND)
            if(fileStateView != null) {
                fileStateView.setText(R.string.ease_have_uploaded);
            }
    }

    protected void setFileIcon(String fileName) {
        EaseFileIconProvider provider = EaseUIKit.getInstance().getFileIconProvider();
        if(provider != null) {
            Drawable icon = provider.getFileIcon(fileName);
            if(icon != null) {
                ivFileIcon.setImageDrawable(icon);
            }
        }
    }

    /**
     * Download file or thumbnail.
     * @param isThumbnail   Whether to download thumbnail
     */
    protected void downloadAttachment(boolean isThumbnail) {
        if(message != null) {
            setMessageDownloadCallback();
            if(isThumbnail) {
                ChatClient.getInstance().chatManager().downloadThumbnail(message);
            }else {
                ChatClient.getInstance().chatManager().downloadAttachment(message);
            }
        }
    }

    @Override
    public void updateView(ChatMessage msg) {
        if(!isDownloading) {
            super.updateView(msg);
        }
    }

    /**
     * Set message download callback.
     */
    protected void setMessageDownloadCallback() {
        if(message != null) {
            isDownloading = true;
            Log.e("message", "setMessageDownloadCallback");
            message.setMessageStatusCallback(new CallBack() {
                @Override
                public void onSuccess() {
                    Log.e("message", "setMessageDownloadCallback onSuccess");
                    post(()->onDownloadAttachmentSuccess());
                    isDownloading = false;
                }

                @Override
                public void onError(int code, String error) {
                    Log.e("message", "setMessageDownloadCallback onError");
                    post(()->onDownloadAttachmentError(code, error));
                    isDownloading = false;
                }

                @Override
                public void onProgress(int progress, String status) {
                    Log.e("message", "setMessageDownloadCallback onProgress");
                    post(()->onDownloadAttachmentProgress(progress));
                }
            });
        }
    }

    protected void onDownloadAttachmentSuccess() {}

    protected void onDownloadAttachmentError(int code, String error) {}

    protected void onDownloadAttachmentProgress(int progress) {}
}

File: ease-im-kit/src/main/kotlin/io/agora/uikit/widget/chatrow/EaseChatRowFile.kt
package io.agora.uikit.widget.chatrow

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import io.agora.uikit.R
import io.agora.uikit.common.ChatClient
import io.agora.uikit.common.ChatDownloadStatus
import io.agora.uikit.common.ChatFileMessageBody
import io.agora.uikit.common.ChatImageMessageBody
import io.agora.uikit.common.ChatNormalFileMessageBody
import io.agora.uikit.common.ChatTextFormater
import io.agora.uikit.common.ChatVideoMessageBody
import io.agora.uikit.common.extensions.isSend
import io.agora.uikit.common.extensions.mainScope
import io.agora.uikit.common.impl.CallbackImpl
import kotlinx.coroutines.launch

/**
 * file for row
 */
open class EaseChatRowFile @JvmOverloads constructor(
    private val context: Context,
    private val attrs: AttributeSet? = null,
    private val defStyleAttr: Int = 0,
    isSender: Boolean = false
) : EaseChatRow(context, attrs, defStyleAttr, isSender) {
    /**
     * file name
     */
    protected val fileNameView: TextView? by lazy { findViewById(R.id.tv_file_name) }

    /**
     * file's size
     */
    protected val fileSizeView: TextView? by lazy { findViewById(R.id.tv_file_size) }

    /**
     * file state
     */
    protected val fileStateView: TextView? by lazy { findViewById(R.id.tv_file_state) }
    private val ivFileIcon: ImageView? by lazy { findViewById(R.id.iv_file_icon) }

    override fun onInflateView() {
        inflater.inflate(
            if (!isSender) R.layout.ease_row_received_file else R.layout.ease_row_sent_file,
            this
        )
    }

    override fun onSetUpView() {
        message?.run {
            (body as? ChatNormalFileMessageBody)?.let {
                fileStateView?.visibility = View.GONE
                fileNameView?.text = it.fileName
                fileSizeView?.text = ChatTextFormater.getDataSize(it.fileSize)
            }
        }
    }

    override fun onMessageSuccess() {
        super.onMessageSuccess()
        message?.run {
            if (isSend()) {
                fileStateView?.setText(R.string.ease_have_uploaded)
            }
        }
    }

    /**
     * Download file or thumbnail.
     * @param isThumbnail   Whether to download thumbnail
     */
    protected fun downloadAttachment(isThumbnail: Boolean) {
        message?.let {
            setMessageDownloadCallback(isThumbnail)
            if (isThumbnail) {
                ChatClient.getInstance().chatManager().downloadThumbnail(it)
            } else {
                ChatClient.getInstance().chatManager().downloadAttachment(it)
            }
        }
    }

    /**
     * Set message download callback.
     */
    protected fun setMessageDownloadCallback(isThumbnail: Boolean) {
        message?.run {
            var canDownload = false
            val body = body
            if (isThumbnail) {
                if (body is ChatImageMessageBody || body is ChatVideoMessageBody) {
                    if (body is ChatImageMessageBody
                        && body.thumbnailDownloadStatus() != ChatDownloadStatus.SUCCESSED) {
                        canDownload = true
                    }
                    if (body is ChatVideoMessageBody
                        && body.thumbnailDownloadStatus() != ChatDownloadStatus.SUCCESSED) {
                        canDownload = true
                    }
                }
            } else {
                if (body is ChatFileMessageBody) {
                    if (body.downloadStatus() != ChatDownloadStatus.SUCCESSED) {
                        canDownload = true
                    }
                }
            }
            if (canDownload) {
                setMessageStatusCallback(CallbackImpl(
                    onSuccess = {
                        context.mainScope().launch {
                            onDownloadAttachmentSuccess()
                        }
                    },
                    onError = { code, error ->
                        context.mainScope().launch {
                            onDownloadAttachmentError(code, error)
                        }
                    },
                    onProgress = { progress ->
                        context.mainScope().launch {
                            onDownloadAttachmentProgress(progress)
                        }
                    }
                ))
            }
        }
    }

    protected open fun onDownloadAttachmentSuccess() {}
    protected open fun onDownloadAttachmentError(code: Int, error: String?) {}
    protected open fun onDownloadAttachmentProgress(progress: Int) {}

    companion object {
        private val TAG = EaseChatRowFile::class.java.simpleName
    }
}